src/lib/Inference.hs

```diff
@@ -935,38 +935,57 @@ inferUDeclTop (UInterface paramBs superclasses methodTys className methodNames)
   extendSubst (className @> className' <.> methodNames @@> methodNames') cont
 inferUDeclTop _ _ = error "not a top decl"
 
-tyConDefAsAtom :: TopInferer m => DataDefName o -> m i o (Atom o)
-tyConDefAsAtom defName = liftBuilder do
-  defName' <- sinkM defName
-  DataDef sourceName params _ <- lookupDataDef defName'
+tyConDefAsAtom :: TopInferer m => DataDefName o -> Maybe (DataIfaceReq o) -> m i o (Atom o)
+tyConDefAsAtom defName maybeIfaceReq = liftBuilder do
+  DataDef sourceName params _ <- lookupDataDef =<< sinkM defName
   buildPureNaryLam (EmptyAbs $ binderNestAsPiNest PlainArrow params) \params' -> do
-    defName'' <- sinkM defName'
-    return $ TypeCon sourceName defName'' $ Var <$> params'
-
-dataConDefAsAtom :: TopInferer m => DataDefName o -> Int -> m i o (Atom o)
-dataConDefAsAtom defName conIx = liftBuilder do
+    EmptyAbs clsBs' <- case maybeIfaceReq of
+      Just clsAbs -> sinkM clsAbs >>= (`applyNaryAbs` params')
+      Nothing     -> return $ EmptyAbs Empty
+    buildPureNaryLam (EmptyAbs $ binderNestAsPiNest ClassArrow clsBs') \_ -> do
```

Comment:
I see, so the interface constraints just become phantom arguments to the type constructor *function*, but they don't appear in the internal data definition itself. I can imagine we might eventually want to change that. For example, we might want the user-defined `Ix n` constraint to be available internally whenever we have something of type `n => a` so we can use it to determine storage requirements. But this is fine for now.

Reply:
Correct, this makes some simplifying assumptions that only work for as long as the type expressions have phantom dependence on type-classes, which is the current situation for `Ix` on my working branch.